text/0016-generalized-entity-store.md

```diff
@@ -0,0 +1,113 @@
+# Generalized Entity Store
+
+## Related issues and PRs
+
+- Implementation PR(s): https://github.com/cedar-policy/cedar/pull/171, which is a pull request from https://github.com/prakol16/cedar
+
+## Timeline
+
+- Start Date: 2023-07-11
+- Date Entered FCP:
+- Date Accepted:
+- Date Landed:
+
+## Summary
+
+Entity storage ideally should not be restricted to JSON files. We should either allow users to add their own custom storage methods or add support for some non-JSON common storage methods (e.g. databases) ourselves.
+
+## Basic example
+
+
+```rust
+struct MyOwnDatabase {
+    // ...
+}
+
+impl EntityDatabase for MyOwnDatabase {
+    fn get_entity_of_uid(&self, uid: &EntityUID) -> Option<Entity> {
+        // custom implementation which could e.g. make a request to a database
+    }
+}
+```
+
+This could be code written by users, or we could hide `EntityDatabase` and instead, there could be several storage solutions (other than the default JSON one) that we write ourselves.
+
+## Motivation
+
+  - Currently, every request reads the entire JSON file, reparsing all of the attributes, even though the overwhelming majority of requests will need to read only a few entries of the entity store. In general, with many entities, it may not even be feasible to load all of the entities into memory at once.
+
+  - In the future, we can turn partial expression evaluations into queries so that we can efficiently find e.g. all of the photos that a user has access to. Ideally, we would leverage existing query languages such as SQL and existing storage solutions which are already optimized to respond to these kinds of queries efficiently, rather than try to rebuild this on our own.
+
+  - Some users may already have infrastructure that uses databases or other storage methods for their applications. This makes integrating Cedar into existing software easier, because users are not boxed in to one particular storage format.
+
+
+## Detailed design
+
+
+We would generalize the authorizer to accept any `EntityDatabase`, not just `Entities`:
+
+```rust
+impl Authorizer {
+    // old:
+    pub fn is_authorized_old(&self, q: &Request, pset: &PolicySet, entities: &Entities) -> Response
+
+    // proposed change:
+    pub fn is_authorized<T : EntityDatabase>(&self, q: &Request, pset: &PolicySet, entities: &T) -> Response
+}
+```
+
+Here is the current interface for `EntityDatabase` (see [the source](https://github.com/prakol16/cedar/blob/e5df52e118e378c2142aaa6d9a4781f0f67b8064/cedar-policy-core/src/entities.rs#L265) for more comments and the implementation of the trait for `Entities`):
+
+```rust
+pub trait EntityDatabase {
+    /// Given a uid, get the corresponding entity if it exists
+    /// The entity should already have an `ancestors` attribute which represents
+    /// the transitive closure ancestors, not the direct parents
+    fn get_entity_of_uid(&self, uid: &EntityUID) -> Option<Entity>;
```

Comment:
We could empower applications to implement their `EntityDatabase` that connects to external database clusters where entity data are stored, if we make this an async method allowing non-blocking io.